Update Analyzer and ensure latest differences data.

hermitdave opened the pull request:
Updating Analyzer for latest SDK versions
Remove Differences for API Contract 6.0.0.0
Add Differences for API Contract 8.0.0.0

Issue: #
https://github.com/microsoft/Rapid-XAML-Toolkit/issues/146#issuecomment-530155254

## PR Type
What kind of change does this PR introduce?
<!-- Please uncomment one ore more that apply to this PR -->

<!-- - Bugfix -->
<!-- - Feature -->
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Sample app changes -->
- Other... Please describe: Update Analyser and consume differences based on latest UWP SDKs


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Older SDKs being referenced

## What is the new behavior?
Update SDKs being referenced

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: <!-- docs PR link -->
- [ ] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [ ] Contains **NO** breaking changes


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected -->


## Other information

michael-hawker commented:
Thanks @hermitdave is the analyzer only able to look at N-1?  Could we have it keep the old differences in order to alert folks of what min they need to target as we'll still be supporting down to 16299 for now?

Also looks like the [CI build is failing](https://dev.azure.com/dotnet/WindowsCommunityToolkit/_build/results?buildId=21189&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=d1e88c09-dceb-53c1-5579-19c846f3526a&l=2603)

```
 CSC : error CS1566: Error reading resource 'Microsoft.Toolkit.Uwp.PlatformSpecificAnalyzer.Differences-6.0.0.0.gz' -- 'Could not find file 'D:\a\1\s\Microsoft.Toolkit.Uwp.PlatformSpecificAnalyzer\Differences-6.0.0.0.gz'.' [D:\a\1\s\Microsoft.Toolkit.Uwp.PlatformSpecificAnalyzer\Microsoft.Toolkit.Uwp.PlatformSpecificAnalyzer.csproj]
```

hermitdave commented:
The N-2 logic is hard code though I can take a look at making it flexible.
Let me check what the issue is with files

michael-hawker commented:
Thanks @hermitdave, let us know if you want to chat and bounce ideas off.  I'm not entirely familiar with how this works, but I remember there was a problem where we didn't know what the min version the application is targeting?  If so, could we use [this type of analyzer example](https://github.com/dotnet/roslyn-sdk/blob/master/samples/CSharp/Analyzers/Analyzers.Implementation/AdditionalFileAnalyzers/SimpleAdditionalFileAnalyzer.cs) to grab info from parsing the csproj manually?

hermitdave commented:
There is no info on how analyzer can read csproj / vbproj files. Info requested from VS / SDK tab fell on deaf ears.

The original analyzer code that this is roughly based on added  dummy file reference when nuget script was run during initial reference and I want comfortable with that mechanism. When project target SDK version is updated, IMO the dummy file wouldn't be updated and old SDK versions will continue to be used by analyzer making it useless. 

I personally do not know of any way.

hermitdave commented:
@michael-hawker I have added support for 16299 and onwards and simplified the logic. I cannot fully the code changes as I cannot compile the toolkit for a long time now having tried both VS 2017 and VS 2019.
Apologies

michael-hawker commented:
@hermitdave did you see @azchohfi comment? Have you been able to verify this and find a solution? Thanks!

hermitdave commented:
@michael-hawker i can't seem to be compile any xaml uwp projects on Surface book. I will try again on a different device but that requires dev setup. As for false positives, the code is dumb - it flags anything that is in changes in the versions supported.

michael-hawker commented:
@azchohfi @hermitdave realizing now that once we move to WinUI 3, it's going to support all its versions, so there's just a single API surface.

Since we're having trouble getting these to work properly, does it instead make sense just to deprecate them and move on? As once we adopt WinUI 3, by construction, it basically solves these types of issues?

Thoughts?

azchohfi commented:
WinUI3 will be RS4+, but this is only for WUX APIs, not everything. This package checks for every WinRT API, not only WUX. I think it still makes sense.

michael-hawker commented:
@hermitdave @azchohfi did they recently add helpers to VS?

![image](https://user-images.githubusercontent.com/24302614/83070576-8b054f00-a020-11ea-8f2f-5bfc8f92884f.png)

azchohfi commented:
No idea.

michael-hawker commented:
@hermitdave since we don't have a good way to know the target of the project and have to continually update diffs but flag everything, should we just work with the VS team to get this added to VS like the XAML analyzers and just drop support for it in the Toolkit?

michael-hawker commented:
@lyalindotcom we've been at a bit of an impasse here, do you know how the [analyzer I saw above works](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/2999#issuecomment-634931127) and who owns it? This is a recently added feature, eh? It'd be great to know what's going on in this space, so if anything we just deprecate our attempts here, remove it from the Toolkit, and leverage/contribute to VS efforts instead? Thanks!

LyalinDotCom commented:
@michael-hawker I'll respond over team

michael-hawker commented:
Forgot we had this discussion already just elsewhere, looping back and closing this issue based on [comment from here](https://github.com/mrlacey/Rapid-XAML-Toolkit/issues/146#issuecomment-637160734).

@azchohfi I've filed issue #3396 for tracking for 7.0.
